ADD ObjectList, Listable, Sorting classes and unit-tests

SvyatoslavZL opened the pull request:
ADD ObjectList - custom analog of ArrayList, Listable interface - analog of List, Sorting class - partial analog of Collections for sorting ObjectList, unit-tests for ObjectList.java, Sorting.java

SvyatoslavZL committed:
ADD ObjectList - custom analog of ArrayList, Listable interface - analog of List, Sorting class - partial analog of Collections for sorting ObjectList, unit-tests for ObjectList.java, Sorting.java

SvyatoslavZL committed:
Merge remote-tracking branch 'origin/task-1' into task-1

SvetaPiven commented on task1/pom.xml:
```diff
@@ -17,4 +17,25 @@
+            <version>5.11.1</version>
+            <scope>test</scope>
+        </dependency>
+<!--        <dependency>-->
```
комментарии такого плана не оставляем. перед пушем или todo или чистим

SvyatoslavZL replied:
Почистил

SvetaPiven commented on task1/src/main/java/org/aston/kovalinsky/ObjectList.java:
```diff
@@ -0,0 +1,206 @@
+     * @param elementToAdd element to be appended to this list
+     */
+    @Override
+    public void add(E elementToAdd) {
```
ты сначала увеличиваешь размер, а только потом проверяешь, достиг ли размер максимума. может привести к ArrayIndexOutOfBoundsException если добавляешь элемент в массив который уже заполнен

SvyatoslavZL replied:
Логика работы add такая: 1) Вставка 2) подготовка под вставку следующего.
Допустим, size = 9. list.add(E e) добавит элемент на 9 индекс, затем увеличит size и сразу проверит совпал ли size с длинной массива. Если да, то увеличит ёмкость для будущего добавления. Тест increaseCapacity это подтверждает.

На самом деле, да, вижу что логичнее было бы сделать наоборот, 

Дополнил описание метода, чтобы объяснить эту логику и добавил тест add, где добаляется больше 10 элементов и не выбрасывается исключение.

SvetaPiven commented on task1/src/main/java/org/aston/kovalinsky/ObjectList.java:
```diff
@@ -0,0 +1,206 @@
+     */
+    @Override
+    public void add(int index, E elementToAdd) {
+        Objects.checkIndex(index, size + 1);
```
Здесь аналогичная проблема см выше

SvyatoslavZL replied:
Добавление элемента по индексу происходит более логично. 1) Увеличивается размер 2) Элементы сдвигаются вправо, чтобы освободить место для нового. 3) Вставляется новый элемент

Дополнил описание метода.

SvetaPiven commented on task1/src/main/java/org/aston/kovalinsky/ObjectList.java:
```diff
@@ -0,0 +1,206 @@
+     */
+    @Override
+    public void remove(int index) {
+        Objects.checkIndex(index, size);
```
после удаления элемента последний остаётся в массиве и не очищается

SvyatoslavZL replied:
Спасибо! Исправил.

SvetaPiven commented on task1/src/main/java/org/aston/kovalinsky/ObjectList.java:
```diff
@@ -0,0 +1,206 @@
+        for (int i = index; i < size - 1; i++) {
+            listData[i] = listData[i + 1];
+        }
+        size--;
```
listData[--size] = null

SvetaPiven commented on task1/src/main/java/org/aston/kovalinsky/ObjectList.java:
```diff
@@ -0,0 +1,206 @@
+     */
+    @Override
+    public void clear() {
+        Arrays.fill(listData, null);
```
можно оптимиз. метод очищая только ту часть массива которая занята элементами

SvyatoslavZL replied:
Спасибо. Сделал

SvetaPiven commented on task1/src/main/java/org/aston/kovalinsky/ObjectList.java:
```diff
@@ -0,0 +1,206 @@
+    }
+
+    private void increaseCapacity() {
+        capacity *= CAPACITY_MULTIPLIER;
```
переменная capacity используется только в методе, и её увеличение никак не влияет на массив

SvetaPiven replied:
можно убрать

SvyatoslavZL replied:
Убрал. Получается, я её сделал, чтобы протестировать увеличение capacity (что показывало работоспособность add) Но, тесты не должны менять под себя класс. Исправлено

SvetaPiven commented on task1/src/main/java/org/aston/kovalinsky/ObjectList.java:
```diff
@@ -0,0 +1,206 @@
+     */
+    @Override
+    @SuppressWarnings("unchecked")
+    public E set(int index, E elementToSet) {
```
можно добавить проверки на null для аргументов если они не подходят

SvyatoslavZL replied:
На вход принимаются либо E - элементы, либо null. Лист может содержать null. Не совсем понял

SvetaPiven commented on task1/src/main/java/org/aston/kovalinsky/Sorting.java:
```diff
@@ -0,0 +1,82 @@
+     * @see Comparable
+     */
+    public static <T extends Comparable<? super T>> void sort(ObjectList<T> list) {
+        quickSort(list, 0, list.size() - 1, Comparator.naturalOrder());
```
NullPointerException

SvyatoslavZL replied:
Если в метод не передать лист или Компаратор, то выйдет NullPointerException. 
Можно поподробнее что нужно исправить?

SvetaPiven commented on task1/src/main/java/org/aston/kovalinsky/Sorting.java:
```diff
@@ -0,0 +1,82 @@
+     * @see Comparator
+     */
+    public static <T> void sort(ObjectList<T> list, Comparator<T> comparator) {
+        quickSort(list, 0, list.size() - 1, comparator);
```
NullPointerException

SvetaPiven commented on task1/src/test/java/org/aston/kovalinsky/ObjectListTest.java:
```diff
@@ -0,0 +1,167 @@
+    @Test
+    void toArray_returnCorrectArray() {
+        addDifferentObjects();
+        Object[] actualArray = testObjectList.toArray();
```
тесты делятся по хорошему на 2-3 части

SvetaPiven replied:
подготовка действие проверка. и лучше разделять пустыми строками это

SvyatoslavZL replied:
Спасибо, подправил

SvyatoslavZL committed:
ADD more detailed description of add method, another add test to ensure that list capacity increases if needed.

FIX remove method duplicated last element and didn't delete it. Clear method was optimized.

DELETE unnecessary dependencies from pom, unnecessary non-final capacity variable from ObjectList, getCapacity method, capacity tests.
